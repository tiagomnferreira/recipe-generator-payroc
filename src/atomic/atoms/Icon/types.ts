import { SVGProps } from "react";
import iconMap from "./iconMap";

export interface Props extends SVGProps<SVGSVGElement> {
  name: keyof typeof iconMap;
}
