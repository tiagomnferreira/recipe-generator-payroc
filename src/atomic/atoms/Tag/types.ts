import { HTMLAttributes } from "react";

export interface Props extends HTMLAttributes<HTMLDivElement> {
  label: string;
  active?: boolean;
}
