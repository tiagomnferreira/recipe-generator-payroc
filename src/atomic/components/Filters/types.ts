import { Ingredient } from "../../../typescript";

export interface Props {}

export interface ContainerProps {
  isOpen: boolean;
}
