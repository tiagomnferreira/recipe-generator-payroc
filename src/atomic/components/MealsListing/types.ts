import { Meal, UnaryFn } from "../../../typescript";

export interface Props {
  meals: Meal[];
}
