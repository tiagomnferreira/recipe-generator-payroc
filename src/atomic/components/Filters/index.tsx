import { useEffect, useState } from "react";
import {
  Container,
  Bar,
  FiltersList,
  Lock,
  Group,
  ClearBtn,
  Divider,
  DividerText,
} from "./styled-components";
import { Props } from "./types";
import Text from "../../atoms/Text";
import Input from "../../atoms/Input";
import { debounce } from "lodash";
import {
  filterBy,
  filterIngredients,
  onChange,
  onSearchIngredient,
  parseIngredientsToLabels,
  parseToLabels,
} from "./utils";
import {
  getAllAreas,
  getAllCategories,
  getAllIngredients,
} from "../../../services";
import TagsListing from "../../molecules/TagsListing";

const Filters = ({ setMeals, getInitialMeals }: Props) => {
  const [isOpen, setIsOpen] = useState(false);
  const [categories, setCategories] = useState<string[]>([]);
  const [ingredients, setIngredients] = useState<string[]>([]);
  const [areas, setAreas] = useState<string[]>([]);
  const [activeTag, setActiveTag] = useState<string | undefined>(undefined);
  const [ingredientFilter, setIngredientFilter] = useState("");

  const toggle = () => setIsOpen((prevState) => !prevState);

  useEffect(() => {
    getAllIngredients().then(({ data }) => {
      setIngredients(parseIngredientsToLabels(data.meals));
    });
    getAllAreas().then(({ data }) => {
      setAreas(parseToLabels(data.meals));
    });
    getAllCategories().then(({ data }) => {
      setCategories(parseToLabels(data.meals));
    });
  }, []);

  return (
    <Container>
      <Bar onClick={toggle}>Filter</Bar>
      <Lock name={isOpen ? "locketOpen" : "locketClosed"} />
      <FiltersList isOpen={isOpen}>
        <Group>
          <Text>By meal name:</Text>
          <Input
            placeholder="Type your meal name"
            onChange={debounce(onChange(setMeals), 1000)}
          />
        </Group>

        <Divider>
          <DividerText>OR</DividerText>
        </Divider>

        <Group>
          <Text>By Category:</Text>
          <TagsListing
            activeTag={activeTag}
            labels={categories}
            onTagClick={(category) =>
              filterBy(setMeals, setActiveTag)("c", category)
            }
          />
        </Group>

        <Divider>
          <DividerText>OR</DividerText>
        </Divider>

        <Group>
          <Text>By Area:</Text>
          <TagsListing
            activeTag={activeTag}
            labels={areas}
            onTagClick={(area) => filterBy(setMeals, setActiveTag)("a", area)}
          />
        </Group>

        <Divider>
          <DividerText>OR</DividerText>
        </Divider>

        <Group>
          <Text>By Ingredient:</Text>
          <Input
            placeholder="Type the ingredient's name"
            onChange={debounce(onSearchIngredient(setIngredientFilter), 1000)}
          />
          <TagsListing
            activeTag={activeTag}
            labels={filterIngredients(ingredientFilter, ingredients)}
            onTagClick={(ingredient) =>
              filterBy(setMeals, setActiveTag)("i", ingredient)
            }
          />
        </Group>
      </FiltersList>
      {isOpen && <ClearBtn onClick={getInitialMeals}>Clear</ClearBtn>}
      <Bar onClick={toggle} />
    </Container>
  );
};

export default Filters;
