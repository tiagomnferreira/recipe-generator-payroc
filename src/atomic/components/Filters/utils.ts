import { ChangeEvent } from "react";
import {
  Area,
  Category,
  FilterType,
  Ingredient,
  Meal,
  NullaryFn,
  SetState,
} from "../../../typescript";
import { getMealByName, getMealsByType } from "../../../services";
import { get } from "lodash";

export const onChange =
  (setMeals: SetState<Meal[]>, getInitialMeals: NullaryFn<void>) =>
  (e: ChangeEvent<HTMLInputElement>) => {
    const name = e.target.value;
    if (name !== "")
      getMealByName(name).then(({ data }) => {
        setMeals(data?.meals || []);
      });
    else getInitialMeals();
  };

export const filterBy =
  (setMeals: SetState<Meal[]>, setActiveTag: SetState<string | undefined>) =>
  (type: FilterType, value: string) => {
    getMealsByType(type, value).then(({ data }) => {
      setMeals(data?.meals || []);
      setActiveTag(value);
    });
  };

export const parseToLabels = <T extends Area | Category>(vec: T[]): string[] =>
  vec
    .map((item: T) => {
      const key = Object.keys(item).pop();

      return get(item, `${key}`, "") as string;
    })
    .filter(Boolean);

export const parseIngredientsToLabels = (ingredients: Ingredient[]) =>
  ingredients.map((item) => get(item, "strIngredient", "")).filter(Boolean);

export const onSearchIngredient =
  (setIngredientFilter: SetState<string>) =>
  (e: ChangeEvent<HTMLInputElement>) =>
    setIngredientFilter(e.target.value);

export const filterIngredients = (filter: string, ingredients: string[]) =>
  ingredients
    .map((item) =>
      item.toLowerCase().includes(filter.toLowerCase()) ? item : ""
    )
    .filter(Boolean)
    .slice(0, 28);
