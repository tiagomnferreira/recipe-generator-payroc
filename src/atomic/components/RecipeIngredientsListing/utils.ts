import { get } from "lodash";
import { Meal } from "../../../typescript";

export const getAllIngredientsFromMeal = (meal: Meal) =>
  Object.keys(meal).reduce((acc: string[], key) => {
    if (key.includes("strIngredient")) {
      const position = key.match(/\d+/)?.pop();
      const ingredient = get(meal, `strIngredient${position}`, "");
      const measurements = get(meal, `strMeasure${position}`, "");

      if (position === undefined || ingredient === "" || measurements === "")
        return acc;

      acc.push(`${ingredient} (${measurements})`);
    }

    return acc;
  }, []);
