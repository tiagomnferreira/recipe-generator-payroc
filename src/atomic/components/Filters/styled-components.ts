import { styled } from "styled-components";
import { COLOR } from "../../../style/palette";
import { generateBoxShadow } from "../../../style/utils";
import { ContainerProps } from "./types";
import Icon from "../../atoms/Icon";
import Button from "../../atoms/Button";
import Text from "../../atoms/Text";

export const Container = styled.section`
  border-radius: 8px;
  display: flex;
  flex-direction: column;
  overflow: hidden;
  box-shadow: ${generateBoxShadow({ blur: 10 })};
  position: fixed;
  bottom: 35px;
  width: 300px;
  background-color: ${COLOR.BACKGROUND};
  z-index: 1;
`;

export const Bar = styled.div`
  background-color: ${COLOR.PRIMARY};
  height: 40px;
  cursor: pointer;
  text-align: center;
  color: ${COLOR.BACKGROUND};
  font-weight: 600;
`;

export const FiltersList = styled.div<ContainerProps>`
  height: ${({ isOpen }) => (isOpen ? "500px" : "0px")};
  padding: ${({ isOpen }) => (isOpen ? "30px 12px 12px 12px" : "0px")};
  transition: all 0.5s ease-in-out;
  position: relative;
  overflow: ${({ isOpen }) => (isOpen ? "auto" : "hidden")};
  border: 1px solid ${COLOR.NEUTRAL};
`;

export const Lock = styled(Icon)`
  position: absolute;
  top: 20px;
  margin: auto;
  left: 130px;
  background-color: ${COLOR.ACCENT};
  border-radius: 20px;
  z-index: 1;
`;

export const Group = styled.div``;

export const ClearBtn = styled(Button)`
  margin: 8px auto;
`;

export const Divider = styled.div`
  border-bottom: 1px solid ${COLOR.NEUTRAL};
  width: 100%;
  height: 0px;
  position: relative;
  margin: 16px 0;
`;

export const DividerText = styled(Text)`
  position: absolute;
  margin: auto;
  left: 45%;
  top: -8px;
  background: white;
  padding: 0 4px;
`;
