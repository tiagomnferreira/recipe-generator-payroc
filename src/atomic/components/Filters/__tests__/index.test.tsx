import { render, screen } from "@testing-library/react";
import { renderWithRouter } from "../../../../utils/testUtils";
import Filters from "..";
import { Props } from "../types";

describe("Filters", () => {
  describe("UI", () => {
    const getInitialMeals = jest.fn();
    const setMeals = jest.fn();

    const props: Props = {
      getInitialMeals,
      setMeals,
    };

    it("should match the generated snapshot", () => {
      const { container } = renderWithRouter(<Filters {...props} />);

      expect(container).toMatchSnapshot();
    });

    it("shoulder render the filters container", () => {
      renderWithRouter(<Filters {...props} />);

      const container = screen.getByRole("listbox", {
        name: "Filters container",
      });

      expect(container).toBeInTheDocument();
    });
  });
});
