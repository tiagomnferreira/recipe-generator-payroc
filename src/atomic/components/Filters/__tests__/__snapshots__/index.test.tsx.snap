// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters UI should match the generated snapshot 1`] = `
<div>
  <section
    aria-label="Filters container"
    class="sc-dlfmnT cMtZwp"
    role="listbox"
  >
    <div
      class="sc-hKgJyM gKjfLd"
    >
      Filter
    </div>
    <svg
      class="sc-jSgtCu iEJvLn"
      fill="none"
      height="40"
      viewBox="0 0 24 24"
      width="40"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M13.275 19H9.151c-1.854-.032-3.33-1.599-3.3-3.5V12c-.017-1.574.997-2.963 2.474-3.39v-.46C8.275 6.463 9.568 5.053 11.213 5c1.645.053 2.937 1.463 2.888 3.15v.46c1.477.427 2.491 1.816 2.475 3.39v3.5c.03 1.901-1.447 3.468-3.3 3.5Z"
        stroke="#000"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
      />
      <path
        d="M10.463 15.5a.75.75 0 0 0 1.5 0h-1.5Zm1.5-1.75a.75.75 0 0 0-1.5 0h1.5Zm1.94-4.416a.75.75 0 0 0 .395-1.448l-.394 1.448Zm-.627-.834v.75-.75Zm-4.125 0v.75-.75Zm-1.022-.614a.75.75 0 1 0 .394 1.448l-.394-1.448Zm3.834 7.614v-1.75h-1.5v1.75h1.5Zm2.335-7.614a3.878 3.878 0 0 0-1.023-.136l.002 1.5c.211 0 .422.028.627.084l.394-1.448Zm-1.022-.136H9.151v1.5h4.125v-1.5Zm-4.124 0c-.346 0-.69.046-1.023.136l.394 1.448c.205-.056.416-.084.627-.084l.002-1.5Z"
        fill="#000"
      />
    </svg>
    <div
      class="sc-eCstFB hiPhNx"
    >
      <div
        class="sc-gKsdJj"
      >
        <p
          class="sc-gsTChe fBXRVx"
          font-size="14"
          fontcolor="#333333"
        >
          By meal name:
        </p>
        <div
          class="sc-jrAHeI kYdNEm"
        >
          <input
            class="sc-kEjaJL ktXoSj"
            placeholder="Type your meal name"
          />
        </div>
      </div>
      <div
        class="sc-fubCSR dOdRZH"
      >
        <p
          class="sc-gsTChe fBXRVx sc-pGawl hsvTho"
          font-size="14"
          fontcolor="#333333"
        >
          OR
        </p>
      </div>
      <div
        class="sc-gKsdJj"
      >
        <p
          class="sc-gsTChe fBXRVx"
          font-size="14"
          fontcolor="#333333"
        >
          By Ingredient:
        </p>
        <div
          class="sc-jrAHeI kYdNEm"
        >
          <input
            class="sc-kEjaJL ktXoSj"
            placeholder="Type the ingredient's name"
          />
        </div>
        <div
          class="sc-crrsTd cMpXGW"
        />
      </div>
      <div
        class="sc-fubCSR dOdRZH"
      >
        <p
          class="sc-gsTChe fBXRVx sc-pGawl hsvTho"
          font-size="14"
          fontcolor="#333333"
        >
          OR
        </p>
      </div>
      <div
        class="sc-gKsdJj"
      >
        <p
          class="sc-gsTChe fBXRVx"
          font-size="14"
          fontcolor="#333333"
        >
          By Category:
        </p>
        <div
          class="sc-crrsTd cMpXGW"
        />
      </div>
      <div
        class="sc-fubCSR dOdRZH"
      >
        <p
          class="sc-gsTChe fBXRVx sc-pGawl hsvTho"
          font-size="14"
          fontcolor="#333333"
        >
          OR
        </p>
      </div>
      <div
        class="sc-gKsdJj"
      >
        <p
          class="sc-gsTChe fBXRVx"
          font-size="14"
          fontcolor="#333333"
        >
          By Area:
        </p>
        <div
          class="sc-crrsTd cMpXGW"
        />
      </div>
    </div>
    <div
      class="sc-hKgJyM gKjfLd"
    />
  </section>
</div>
`;
